hail/src/main/scala/is/hail/expr/ir/Simplify.scala

```diff
@@ -319,8 +319,7 @@ object Simplify {
       TableFilter(t,
         ApplySpecial("&&", Array(p1, p2)))
 
-    case TableOrderBy(child, sortFields) if sortFields.isEmpty =>
-      child
+    case TableOrderBy(TableKeyBy(child, _, _), sortFields) => TableOrderBy(child, sortFields)
```

Comment:
I don't think this is general enough. What about adding:
```scala
case TableOrderBy(child, sortFields)
  if TableOrderBy.isAlreadyOrdered(sortFields, child.rvdType.key) =>
  TableKeyBy(child, Array(), false)
case TableKeyBy(TableKeyBy(child, sortFields, false), IndexedSeq(), _) =>
  TableOrderBy(child, sortFields)
```

Reply:
The spec in the google doc wouldn't allow for either of these rewrites. Since we can rewrite a `TableKeyBy(TableKeyBy(child, _), newKey)` as `TableKeyBy(child, newKey)`, the first would lead to optimization totally blowing away the order. We can't remove TableOrderBy nodes, even if a KeyBy substitution in-place may have the same semantics.

The latter is also a deoptimization - keying by an empty key doesn't guarantee a stable sort, so we don't actually have to do the inner keyBy at all.

Reply:
Yup, you're right. This change makes me a little uncomfortable, because the interaction between `TableOrdeBy` and `TableKeyBy` is now more complicated. I was trying to find a normalizing set of rewrite rules to handle that interaction (the second rule was only to make it confluent). I'll approve for now but keep thinking about it.